pkg/runtime/controller.go

```diff
@@ -0,0 +1,424 @@
+	var timeChan <-chan time.Time
+	var timer *time.Timer
+	events := make([]*store.Event, 0, maxEvents)
+
+	for {
+		select {
+		case ev := <-wch:
+			if len(events) == 0 {
+				timer = time.NewTimer(watchFlushDuration)
+				timeChan = timer.C
+			}
+			events = append(events, &ev)
+		case <-timeChan:
+			timer.Stop()
+			timeChan = nil
+			glog.Infof("Publishing %d events", len(events))
+			applyEvents(events)
+			events = events[:0]
+		}
+	}
+}
+
+// applyEvents applies given events to config state and then publishes a snapshot.
+func (c *Controller) applyEvents(events []*store.Event) {
+	for _, ev := range events {
+		switch ev.Type {
+		case store.Update:
+			c.configState[ev.Key] = ev.Value
+		case store.Delete:
+			delete(c.configState, ev.Key)
+		}
+	}
+	c.publishSnapShot()
+}
+
+// validInstanceConfigs returns instanceConfigs from the configState that
+// point to valid templates.
+func (c *Controller) validInstanceConfigs() map[string]*cpb.Instance {
+	instanceConfig := make(map[string]*cpb.Instance)
+
+	// first pass get all the validated instance and handler references
+	for k, cfg := range c.configState {
+		if _, found := c.templateInfo[k.Kind]; !found {
+			continue
+		}
+		instanceConfig[k.String()] = &cpb.Instance{
+			Name:     k.Name,
+			Template: k.Kind,
+			Params:   cfg,
+		}
+	}
+	return instanceConfig
+}
+
+// validHandlerConfigs returns handlerConfigs from the configState that
+// point to valid adapters.
+func (c *Controller) validHandlerConfigs() map[string]*cpb.Handler {
+	handlerConfig := make(map[string]*cpb.Handler)
+	for k, cfg := range c.configState {
+		if _, found := c.adapterInfo[k.Kind]; !found {
+			continue
+		}
+		handlerConfig[k.String()] = &cpb.Handler{
+			Name:    k.Name,
+			Adapter: k.Kind,
+			Params:  cfg,
+		}
+	}
+	return handlerConfig
+}
+
+type rulesByName map[string]*Rule
+
+// rulesByName indexed by namespace
+type rulesMapByNamespace map[string]rulesByName
+
+// rulesListByNamespace is the type needed by resolver.
+type rulesListByNamespace map[string][]*Rule
+
+// convertToRuntimeRules converts internal rules to the format that resolver needs.
+func convertToRuntimeRules(ruleConfig rulesMapByNamespace) (rulesListByNamespace, int) {
+	// convert rules
+	nrules := 0
+	rules := make(rulesListByNamespace)
+	for ns, nsmap := range ruleConfig {
+		rulesArr := make([]*Rule, 0, len(nsmap))
+		for _, rule := range nsmap {
+			rulesArr = append(rulesArr, rule)
+		}
+		rules[ns] = rulesArr
+		nrules += len(rulesArr)
+	}
+	return rules, nrules
+}
+
+// processRules builds the current consistent view of the rules keyed by Namespace and then Name.
+// ht (handlerTable) keeps track of handler-instance association.
+func (c *Controller) processRules(handlerConfig map[string]*cpb.Handler,
+	instanceConfig map[string]*cpb.Instance, ht *handlerTable) rulesMapByNamespace {
+	// current consistent view of the rules
+	// keyed by Namespace and then Name.
+	ruleConfig := make(rulesMapByNamespace)
```

Comment:
Config to me seems more of a proto object... here this is a fully resolved Rule instance..
Suggestion : Renamed method to getResolvedRules and this variable to resolvedRules